clang/include/clang/AST/DeclCXX.h

```diff
@@ -2007,6 +2022,21 @@ class CXXDeductionGuideDecl : public FunctionDecl {
   /// this is an implicit deduction guide.
   CXXConstructorDecl *getCorrespondingConstructor() const { return Ctor; }
 
+  /// Get the deduction guide from which this deduction guide was generated,
+  /// if it was generated as part of alias template deduction or from an
+  /// inherited constructor.
+  CXXDeductionGuideDecl *getSourceDeductionGuide() const {
+    return SourceDeductionGuide.getPointer();
+  }
+
+  void setSourceDeductionGuide(CXXDeductionGuideDecl *DG) {
+    SourceDeductionGuide.setPointer(DG);
+  }
+
+  SourceKind getSourceKind() const { return SourceDeductionGuide.getInt(); }
```

Comment:
Since this 'kind' is meaningless in case SourceDeductionGuide is not set, should we assert here, or perhaps add a third kind, 'None'?

Reply:
I've opted for adding an additional kind `None` for this case, I think it makes the code cleaner by not having `Alias` serve as the default and needing to check the pointer before being able to read the kind.